### ISSUE TITLE
Failed source line for multiline compound expecatations is sometimes confusing

### ISSUE THREAD

**Issue by yujinakayama** [rspec/rspec-expectations#790]
Here's multiline compound expecations:

```ruby
describe 'confusing failed source line with multiline compound expectations' do
  before do
    @foo = 1
    @bar = 1
  end

  def do_something
    @foo += 1
  end

  it 'always shows the first line of the expectation regardless of actual failed line' do
    expect { do_something }.to change { @foo }
                          .and change { @bar }
  end
end
```

and run:

```
Failures:

  1) confusing failed line with multiline compound expectations always shows the first line of the expectation regardless of actual failed line
     Failure/Error: expect { do_something }.to change { @foo }
       expected result to have changed, but is still 1
     # ./spec/example_spec.rb:12:in `block (2 levels) in <top (required)>'
```

The failure message is seemingly saying the `change { @foo }` is failed, but actually it's not. I know this is because the failed line is simply read from source file based on caller location without parsing the Ruby code.

* https://github.com/rspec/rspec-core/blob/v3.2.3/lib/rspec/core/notifications.rb#L243-L267

Currently I'm not sure how this should be solved. Maybe it can be improved by parsing the source as Ruby code but I think it's overkill a bit.

Thoughts?

**Comment by myronmarston** [comment]
I've actually been thinking about the source we display, too.  Even for non-compounds, it's not always ideal.  For example, if you spread an expectation across multiple lines:

``` ruby
expect {
  some_long_expression
}.to change { x }.by(1)
```

...then the failure only prints `expect {` since that's the location of `expect`.  I've been meaning to bring up this issue with you, @yujinakayama, since I figured you've dealt with source issues a lot in transpec :).

> Currently I'm not sure how this should be solved. Maybe it can be improved by parsing the source as Ruby code but I think it's overkill a bit.

Parsing the source came to my mind, too.  Can it be done without introducing additional dependencies?

Another idea that just popped in my head: perhaps matchers can keep track of their instantiation location,  and we can embed the matcher in `ExpectationNotMetError` (as discussed in #694).  Then rspec-core can print all lines from the `expect` location to the matcher location to (hopefully) get the entire expression.

One downside: it's probably less accurate than source parsing (particularly for matchers that may have chained fluent interfaces that spread on to later lines) and `caller`/`caller_locations` is an expensive call....doing this would require using `caller`/`caller_locations` for _every_ matcher instantiation, and not just the failing ones that will benefit from it.

**Comment by myronmarston** [comment]
I think I've thought of a workable way to do this.

* Use stdlib's `ripper` on rubies that provide it as a parser.  (On rubies that don't provide it, we'd fall back to the current behavior).
* Use ripper to figure out how many lines to excerpt.  Start with a single line as it is now and see if that parses. If so, use it.  If not, add the next line and parse that.  If it parses, use that.  If not, add another and repeat.
* Once we've used ripper to figure out which lines form a complete parsable ruby expression, extract those lines and print all of them.  I imagine we'd have some limit so that if we can't find a parsable expression after N lines we'd give up and just print one line now.
* For bonus points, use coderay if available to syntax highlight the snippet.  (We already use it if available for the HTML formatter).

This isn't a foolproof solution, but it seems simple enough and should work for most cases.  Thoughts?

**Comment by JonRowe** [comment]
I'm mildly concerned about requiring things from the std lib (as usual)...

**Comment by yujinakayama** [comment]
> * Use ripper to figure out how many lines to excerpt. Start with a single line as it is now and see if that parses. If so, use it. If not, add the next line and parse that. If it parses, use that. If not, add another and repeat.

This would parse the lines wrongly if the extracted first line is valid Ruby but actually part of multiline expression:

```ruby
    expect { do_something }.to change { @foo }
                          .and change { @bar }
```

Also, I think parsing the source multiple times might be less effective than parsing entire source only once.

Actually I've been using [Parser](https://github.com/whitequark/parser) for Ruby source analysis so I'm not sure if we can do it with Ripper because it gives us less information than Parser. I think I'll try implementing a prototype and see how it's going this weekend.

**Comment by yujinakayama** [comment]
@JonRowe I think we can require ripper lazily when failed so it won't affect users' project at least when everything is green.

**Comment by myronmarston** [comment]
> @JonRowe I think we can require ripper lazily when failed so it won't affect users' project at least when everything is green.

:+1:  Even apart from this, I think it would be a worthwhile tradeoff.

Have you had a chance to prototype this @yujinakayama?

**Comment by myronmarston** [comment]
Closing in favor of rspec/rspec-core#2083.
